### PR TITLE
redux-persist doc

### PR DESCRIPTION
**Pull request by chapati23**
first draft to using redux-persist and the cookies storage adapter together with redux-first-router

### CODE REVIEW

**Comment by faceyspacey** [comment]
question: so do we not have the issue that @ganlub mentioned regarding delaying actions until the store is rehydrated from redux-persist?

If we do, another option is to use the `initialDispatch: false` [option](https://github.com/faceyspacey/redux-first-router/blob/master/docs/connectRoutes.md#options) to get access to an `initialDispatch()` function we can call when we want to. 

That might only solve things for RFR and not other enhancers. If you're not familiar with what I'm talkin about, I actually describe it in step #10 in the latest RFR server-rendering article that came out on friday:

- [Server-Render like a Pro /w Redux-First Router in 10 steps](https://medium.com/faceyspacey/server-render-like-a-pro-w-redux-first-router-in-10-steps-b27dd93859de)

**Comment by chapati23** [comment]
I did not run into this issue, no.

**Comment by MrLoh** [comment]
I'm just trying this on React native and I had the issue where I need the delay.

**Comment by MrLoh** [comment]
The initial route action fires before rehydration has completed and thus won't restore the previous route anymore. I guess on web you'd kind of want this. But not on native.

**Comment by chapati23** [comment]
@gnalub / @MrLoh care to add a `React Native` section to this doc with guidance on delaying / buffering actions until `REHYDRATE` has fired?

**Comment by MrLoh** [comment]
The manual initial dispatch definitely sounds like the easiest option. But I'm actually not sure how persistence in react native should behave. There are two cases:

1. App is opened from home screen after being closed normally: 
    - expected behavior: Start in previous route (maybe don't restore the whole history though, at least twitter doesn't do a full history restore) I guess it's a little confusing when you haven't used the app in a while and then end up in a deeply nested stack or something when you restart. Not sure about this though. 
2. You open the app through a link on the web or from a push notification:
    - expected behavior: you start at the provided route, potentially you need to initialize a stack router so the go back button will function properly. This might override an existing stack. 

So how would that be implemented and what all should be supported. I guess if the whole route history should be restored, the history object would also need to be created from memory. But this seems like overkill maybe, not sure what must be in the history for a stack router to enable going back properly. 

Probably all of this is a bit to complex to include in this pull request and just the option to dispatch the initial route manually should be mentioned and there should be a reference about how to initialize properly on React native based on scenario 1 and 2 with persistence.

**Comment by faceyspacey** [comment]
regarding the StackNavigator and history--basically I see it like this: if you deep-link, u probably should wipe previous history state, but if not, just always remember exactly where the user was. I much prefer that experience over starting over again like most apps do.

Regarding the overkill stuff, the  history state plus memory memoryHistory is basically an object with an array. So the only overkill is the views like StackNavigator itself, which you need anyway.

Otherwise I think in the React Navigation docs I discuss how to do back handling--basically drown the routes of each of your Stack Routers (i.e. stack navigators). That's the most fool proof way to do it. There is some manual work--and it's a good thing--it allows u to configure logic on what routes should be drained first and in what order.

...actually it looks like i dont discuss it. But basically in the handler u use for your BackHandler, dispatch a React Navigation action with the nav key of the corresponding StackRouter you want to drain, and intertwine such actions with actions to close drawers, and shift tabs. Shifting tabs will likely happen from backing out of the stack router attached to it. I gotta pattern for this that is super nice. I wish I documented it. crap.

**Comment by MrLoh** [comment]
Yeah the thing with the stack routers is a bit confusing, I build my own logic Ina thunk, the hat gets the current active tab and then decides what to do, basically just enabling to go back in stack routers and nothing else. That is fine. 

What is also missing to be documented is how to initialize a stack router together with the history properly. Basically we don't just need to dispatch the initial route later, but get the initial route / stack configuration from the persisted state and then dispatch that to kick off the app in the previous state. 

I don't know how history works, so I'm not sure what needs to be done there, does the history live separately from your redux state. The stack router states can simply be restored as everything else as they are just part of the redux state.

**Comment by MrLoh** [comment]
I think the whole interaction between the history and the stack routers, which also contain their history is a bit confusing in react native as those seem like similar concepts with slight differences that you don't really need both. But my understanding here is very limited still.

**Comment by faceyspacey** [comment]
memoryHistory lives separately -> then you pass it to redux-first router -> then reducers get it -> then stack routers get it.

the main thing from the React Navigation boilerplate (though I turned it off at the top) is that you can wait on the promise with the incoming URL and then render your app once you get that. rather than setting up redux only to have to wait 50 more ms before the incoming URL promise resolves. this avoids unnecessary dispatches.

Basically u dont render anything until the incoming URL and router/redux is setup. Then stacknavigators/routers render from state like nothing just happened. It's tested in iOS. 

I didnt get to test Android whether waiting on the incoming URL promise is possible. It wasn't possible in older versions of React Native. but it became possible in 44.

**Comment by faceyspacey** [comment]
bottom line is it's now seamless. u setup ur redux store as shown in the redux first router navigation boilerplate (you uncomment the code that waits on the incoming url promise), then u setup redux, then u render ur app for the first time.

**Comment by faceyspacey** [comment]
u may lose 50 ms before u show something on screen, but not really, since ur showing an overlay otherwise, and since u wouldnt render in a complete state otherwise.

**Comment by MrLoh** [comment]
So this is all for case 2 where my initial route comes from the URL. As for case 1 I'm still not sure. The history needs to be persisted to asyncStorage somehow I guess and then be restored from async storage similar to how we get an incoming URL otherwise, so the app can be kicked off in the last active route with the full history restored. 

But then should the location state be persisted with redux persist or rather not, because it is created from the restored history.

**Comment by MrLoh** [comment]
Not sure how history works. But from my understanding it lives separately and is kind of duplicated in the redux store. I'm not really sure what I need the history object for on React Native. What would I lose if I didn't have it?

**Comment by faceyspacey** [comment]
check the history package docs. they're short. basically u do: `const history = createMemoryHistory()` and now u have the object, and then once u pass it to RFR, u no longer worry about it.

I think the answer ur looking for is this:

when using persist, u will already have a history, and so it might break RFR which needs a *separate* history object to start. in short, and unfortunately, it's sounding like we must finish hydrating from persist, then grab the previous history and put it into the external `history` object which RFR will continue to use under the hood. 

this may require some more work under the hood to automate this unfortunately. at least for Native. Cuz here's the thing: web likely worked cuz there is no history.entries array, as the browser doesn't maintain it for u (or u would have access to other people's history). So that's why people got it to work on web. But on native, we are making use of the history so u can back out of ur entire history.entries. In which case we have the problem i just described. It's solveable, tho likely not supported yet.

I'd build this without persist until the end. I need help by the way. If u wanna add it sooner than later I can make it painless. Otherwise, it's just a little while longer until the RN revisit as you know. 

But thanks to bringing this to my attention. This is a problem i gotta solve. I used to use redux-storage, but when i built RFR didnt use storage (which is similar to persist). So i wasnt aware of these problems.

----
to implement this, we basically need RFR to wait internally until persist is loaded (perhaps by listening directly to its dispatch actions). Then when we receive it, we push the entries on to the internal secondary memory history object we sync to.

Lastly, in userland, we need a way to signal when we're doing deep linking and when we're not. If we are doing deep-linking, we should just clear the history entries in both persist/redux + the memory history, and just start fresh. It doesnt need to be that way though. That's probably just what's most commonly expected. It could also just function like any other new route visited: it's pushed on the end of the history entries array.

**Comment by MrLoh** [comment]
Yeah exactly. That's what I though that in Native this will become more complicated. Of course on web it's all different since you don't expect a website app to start anywhere else than where you got with the link. 

It might be an idea to build persistence into RFR for native independent of redux persist. All you need is a history object that is stored to async storage and optionally retrieved on app start. Maybe building an AsyncStorageHistory is more towards what you actually want. 

But I guess the main challenge is exactly how to persist both the history object and the store and how to correctly restore both to the former state upon starting your app.

**Comment by faceyspacey** [comment]
redux-persist is the only game in town, so coupling to them may be a fine option. redux-storage died a while back when its maintainer left it.

**Comment by faceyspacey** [comment]
Thanks for this!
